modules/http4s-kernel/src/smithy4s/http4s/kernel/RequestDecoder.scala

```diff
@@ -16,8 +16,133 @@
 
 package smithy4s.http4s.kernel
 
+import cats.MonadThrow
+import cats.effect.Concurrent
+import cats.syntax.all._
+import org.http4s.EntityDecoder
 import org.http4s.Request
+import smithy4s.PartialData
+import smithy4s.http.HttpRestSchema
+import smithy4s.http.Metadata
+import smithy4s.schema._
 
 trait RequestDecoder[F[_], A] {
   def decodeRequest(request: Request[F]): F[A]
 }
+
+object RequestDecoder {
```

Comment:
to @Baccata 

I'm not sure I'll be able to remove the deduplication between this and the `ResponseDecoder`. The duplication makes the code simple, I feel like introducing a way to abstract / compose might get in the way of readability more than anything

happy to be proven wrong though

Reply:
@daddykotex, here's some de-dup for the request/response encoders, that re-uses existing constructs to compose things : 

https://github.com/disneystreaming/smithy4s/pull/1023/commits/0545dab6c2467020deeb2295d17255d6d39e9142